subprojects/build-cache/src/main/java/org/gradle/caching/internal/DefaultBuildCacheHasher.java

```diff
@@ -97,8 +98,13 @@ public DefaultBuildCacheHasher putNull() {
         return this;
     }
 
+    @Override
+    public void markAsInvalid() {
+        this.valid = false;
+    }
+
```

Comment:
To conserve memory we could check `valid == true` in each of the `put*()` methods.

Reply:
I wouldn't optimize for the case when we are in a broken state.

Reply:
I don't necessarily agree that it's a broken state, but I also don't feel strongly enough about this.